protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,739 @@
+                                    // was successfully contacted or not. See
+                                    // [`ClosestDisjointPeersIter::on_success`] for details.
+                                    ResponseState::Waiting => {},
+                                    ResponseState::Succeeded => {
+                                        // Given that iterator was not the first to contact the peer
+                                        // it will not be made aware of the closer peers discovered
+                                        // to uphold the S/Kademlia disjoint paths guarantee. See
+                                        // [`ClosestDisjointPeersIter::on_success`] for details.
+                                        iter.on_success(&peer, std::iter::empty());
+                                    },
+                                    ResponseState::Failed => {
+                                        iter.on_failure(&peer);
+                                    },
+                                }
+                            },
+                            None => {
+                                // The iterator is the first to contact this peer.
+                                self.contacted_peers.insert(
+                                    peer.clone().into_owned(),
+                                    PeerState::new(i),
+                                );
+                                return PeersIterState::Waiting(Some(Cow::Owned(peer.into_owned())));
+                            },
+                        }
+                    }
+                    PeersIterState::WaitingAtCapacity => {
+                        match state {
+                            Some(PeersIterState::Waiting(Some(_))) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            Some(PeersIterState::Waiting(None)) => {}
+                            Some(PeersIterState::WaitingAtCapacity) => {}
+                            Some(PeersIterState::Finished) => {
+                                // `state` is never set to `Finished`.
+                                unreachable!();
+                            },
+                            None => state = Some(PeersIterState::WaitingAtCapacity),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Finished => break,
+                }
+            }
+        }
+
+        state.unwrap_or(PeersIterState::Finished)
+    }
+
+    pub fn finish(&mut self) {
+        for iter in &mut self.iters {
+            iter.finish()
+        }
+    }
+
+    pub fn into_result(self) -> impl Iterator<Item = PeerId> {
```

Comment:
Do I understand correctly that this implementation picks one peer from each iterator to go into the result, stopping on `num_results`? Why do you want to cut off the results in this manner? It would also be good to preserve the order from closest to furthest to the target, which is what the other iterators return. Admittedly I was just thinking that we would return a (custom) `Iterator` impl here that draws elements from the underlying iterators on demand. Since all the underlying iterators give an ordered result, this should be doable with a minimum of intermediate buffering (i.e. we're merging `n` ordered sequences into a single ordered sequence, which can be done in a "streaming" / "lazy" fashion since we always just need to look at the next element of each iterator to decide which one is "smallest" and must be returned, thus buffering at most `n-1` elements at a time in this combined iterator).

Reply:
Your suggestion sounds good to me. Sorry for ignoring it earlier.

Just to make sure, does it matter whether `GetClosestPeers` or `GetProviders` return more than `num_results` peers?

```rust
            QueryInfo::GetClosestPeers { key, .. } => {
                Some(KademliaEvent::GetClosestPeersResult(Ok(
                    GetClosestPeersOk { key, peers: result.peers.collect() }
                )))
            }

            QueryInfo::GetProviders { key, providers } => {
                Some(KademliaEvent::GetProvidersResult(Ok(
                    GetProvidersOk {
                        key,
                        providers,
                        closest_peers: result.peers.collect()
                    }
                )))
            }
```

It would not be save for a user to just take the *x* first nodes and ignore the rest, as those *x* might not include all paths.

Reply:
> Just to make sure, does it matter whether GetClosestPeers or GetProviders return more than num_results peers?

`get_providers` and `get_closest_peers` say nothing about the number of results and it can currently not even be influenced with a quorum. `get_record` has the quorum, but a quorum always just gives a lower bound. It is up to the user how many of these results are being "used". There is nothing inherently unsafe in not using all results for the user and discarding results is always something that cannot be expected to be irrelevant, i.e. you need to know what you are doing. The important part to me is only that the subsequent store operations that internally follow a `GetClosestPeers` use all peers.

Reply:
This general approach of doing round-robin is not correct, as I explained [here](https://github.com/libp2p/rust-libp2p/pull/1473#discussion_r423338613).